### PR TITLE
schema_storage: fix schema GC threshold & improve memory management

### PR DESCRIPTION
**Pull request by liuzix**
### What problem does this PR solve? 
- Memory consumption does not fall back after processing a large number of DDLs. #3174 

### What is changed and how it works?
- Correct the calculation of the threshold used to GC schemas.
- Optimize the slice operations used when GC'ing the schemas.

### Check List 

Tests 

 - Unit test
 - Integration test

Related changes

 - Need to cherry-pick to the release branch

### Release note 

```release-note
Fix memory leak after processing DDLs.
```

### CODE REVIEW

File: cdc/processor/processor.go [hunk 1]
```diff
@@ -776,11 +762,9 @@ func (p *processor) doGCSchemaStorage() {
 		// schemaStorage is nil only in test
 		return
 	}
-	// Delay GC to accommodate pullers starting from a startTs that's too small
-	// TODO fix startTs problem and remove GC delay, or use other mechanism that prevents the problem deterministically
-	gcTime := oracle.GetTimeFromTS(p.changefeed.Status.CheckpointTs).Add(-schemaStorageGCLag)
-	gcTs := oracle.ComposeTS(gcTime.Unix(), 0)
-	p.schemaStorage.DoGC(gcTs)
```
Comment (amyangfei) [inline cdc/processor/processor.go:783]: What is the `20min` GC delay used for originally?

Comment (liuzix) [reply cdc/processor/processor.go:783]: It was used to accommodate some buggy behaviors in the old processor, where a puller is started from a timestamp less than the global checkpoint. The new owner does not have this problem.

File: cdc/processor/processor.go [hunk 2]
```diff
@@ -771,16 +761,23 @@ func (p *processor) removeTable(table tablepipeline.TablePipeline, tableID model
...
+	// Please refer to `unmarshalAndMountRowChanged` in cdc/entry/mounter.go
+	// for why we need -1.
+	lastSchemaTs := p.schemaStorage.DoGC(p.changefeed.Status.CheckpointTs - 1)
+	if p.lastSchemaTs == lastSchemaTs {
```
Comment (amyangfei) [inline cdc/processor/processor.go:772]: `p.lastSchemaTs` is not set when updated.

File: metrics/grafana/ticdc.json [hunk 3]
```diff
@@ -3089,6 +3089,126 @@
...
+          },
+          "pluginVersion": "7.5.7",
+          "targets": [
+            {
+              "exemplar": true,
+              "expr": "ticdc_processor_gc_ts{changefeed=~\"$changefeed\", capture=~\"$capture\"}",
+              "format": "table",
+              "interval": "",
+              "legendFormat": "",
+              "queryType": "randomWalk",
```
Comment (overvenus) [inline metrics/grafana/ticdc.json:3171]: randomWalk? It may display random data.

Comment (liuzix) [reply metrics/grafana/ticdc.json:3172]: I think it was added automatically and has no impact on the data.

File: cdc/processor/metrics.go [hunk 4]
```diff
@@ -60,6 +60,13 @@ var (
 			Name:      "exit_with_error_count",
 			Help:      "counter for processor exits with error",
 		}, []string{"changefeed", "capture"})
+	processorSchemaStorageGcTsGauge = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "ticdc",
+			Subsystem: "processor",
+			Name:      "gc_ts",
```
Comment (overvenus) [inline cdc/processor/metrics.go:67]: Could you add a prefix?
```suggestion
			Name:      "schema_storage_gc_ts",
```

File: cdc/processor/processor.go [hunk 5]
```diff
@@ -771,16 +761,24 @@ func (p *processor) removeTable(table tablepipeline.TablePipeline, tableID model
...
 		return
 	}
-	// Delay GC to accommodate pullers starting from a startTs that's too small
-	// TODO fix startTs problem and remove GC delay, or use other mechanism that prevents the problem deterministically
-	gcTime := oracle.GetTimeFromTS(p.changefeed.Status.CheckpointTs).Add(-schemaStorageGCLag)
-	gcTs := oracle.ComposeTS(gcTime.Unix(), 0)
-	p.schemaStorage.DoGC(gcTs)
+	// Please refer to `unmarshalAndMountRowChanged` in cdc/entry/mounter.go
+	// for why we need -1.
+	lastSchemaTs := p.schemaStorage.DoGC(p.changefeed.Status.CheckpointTs - 1)
```
Comment (overvenus) [inline cdc/processor/processor.go:771]: Will `lastSchemaTs` regress?

**Review by overvenus — COMMENTED** [review]
Rest LGTM

File: cdc/processor/processor.go (hunk 5 above)
Comment (liuzix) [reply cdc/processor/processor.go:771]: No. It would have panicked.

File: metrics/grafana/ticdc.json [hunk 6]
```diff
@@ -3089,6 +3089,127 @@
...
+          "options": {
+            "showHeader": true,
+            "sortBy": [
+              {
+                "desc": true,
+                "displayName": "last-snap-ts"
+              }
+            ]
+          },
+          "pluginVersion": "7.5.7",
```
Comment (amyangfei) [inline metrics/grafana/ticdc.json:3163]: release branch still uses `6.1.6`, can we remove this field

**Review by amyangfei — APPROVED** [review]

**Review by overvenus — APPROVED** [review]

**Pull request merged**